src/semidiscretization/semidiscretization_coupled_p4est.jl

```diff
@@ -0,0 +1,544 @@
+    for i in eachsystem(semi_coupled)
+        cb = analysis_callback_coupled.callbacks[i]
+        semi = semi_coupled.semis[i]
+        u_ode = get_system_u_ode(u_ode_coupled, i, semi_coupled)
+        du_ode = get_system_u_ode(du_ode_coupled, i, semi_coupled)
+        initialize!(cb, u_ode, du_ode, t, integrator, semi)
+    end
+end
+
+# This method gets called from OrdinaryDiffEq's `solve(...)`
+function (analysis_callback_coupled::AnalysisCallbackCoupled)(integrator)
+    semi_coupled = integrator.p
+    u_ode_coupled = integrator.u
+    du_ode_coupled = first(get_tmp_cache(integrator))
+
+    # Loop over coupled systems' callbacks and call them individually
+    for i in eachsystem(semi_coupled)
+        @unpack condition = analysis_callback_coupled.callbacks[i]
+        analysis_callback = analysis_callback_coupled.callbacks[i].affect!
+        u_ode = get_system_u_ode(u_ode_coupled, i, semi_coupled)
+
+        # Check condition and skip callback if it is not yet its turn
+        if !condition(u_ode, integrator.t, integrator)
+            continue
+        end
+
+        semi = semi_coupled.semis[i]
+        du_ode = get_system_u_ode(du_ode_coupled, i, semi_coupled)
+        analysis_callback(u_ode, du_ode, integrator, semi)
+    end
+end
+
+# used for error checks and EOC analysis
+function (cb::DiscreteCallback{Condition, Affect!})(sol) where {Condition,
+                                                                Affect! <:
+                                                                AnalysisCallbackCoupled}
+    semi_coupled = sol.prob.p
+    u_ode_coupled = sol.u[end]
+    @unpack callbacks = cb.affect!
+
+    uEltype = real(semi_coupled)
+    l2_error_collection = uEltype[]
+    linf_error_collection = uEltype[]
+    for i in eachsystem(semi_coupled)
+        analysis_callback = callbacks[i].affect!
+        @unpack analyzer = analysis_callback
+        cache_analysis = analysis_callback.cache
+
+        semi = semi_coupled.semis[i]
+        u_ode = get_system_u_ode(u_ode_coupled, i, semi_coupled)
+
+        l2_error, linf_error = calc_error_norms(u_ode, sol.t[end], analyzer, semi,
+                                                cache_analysis)
+        append!(l2_error_collection, l2_error)
+        append!(linf_error_collection, linf_error)
+    end
+
+    (; l2 = l2_error_collection, linf = linf_error_collection)
+end
+
+################################################################################
+### SaveSolutionCallback
+################################################################################
+
+# Save mesh for a coupled semidiscretization, which contains multiple meshes internally
+function save_mesh(semi::SemidiscretizationCoupledP4est, output_directory, timestep = 0)
+    for i in eachsystem(semi)
+        mesh, _, _, _ = mesh_equations_solver_cache(semi.semis[i])
+
+        if mesh.unsaved_changes
+            mesh.current_filename = save_mesh_file(mesh, output_directory; system = i,
```

Comment:
Isn't this potentially type unstable ? `system` has default string, right? So maybe better convert `i` to string here?